packages/lit-dev-cloudbuild-cleanup/src/cleanup.ts

```diff
@@ -0,0 +1,272 @@
+    for (const pr of prs) {
+      openPrs.add(pr.number);
+    }
+  }
+
+  console.log('found open prs', openPrs);
+
+  const run = google.run('v1');
+
+  const auth = new google.auth.GoogleAuth({
+    scopes: ['https://www.googleapis.com/auth/cloud-platform'],
+  });
+
+  const authClient = await auth.getClient();
+  google.options({auth: authClient});
+
+  // fetch list of revisions across both services
+  console.log('fetching list of revisions');
+  const {data} = await run.projects.locations.revisions.list({
+    parent: `projects/${PROJECT_ID}/locations/${_DEPLOY_REGION}`,
+  });
+
+  let revisions = data;
+
+  if (!revisions.items) {
+    throw new Error('found no revision items');
+  }
+
+  console.log('total revisions count', revisions.items.length);
+
+  const revisionsToKeepByRecency = new Set<string>();
+  for (const rev of revisions.items) {
+    if (!rev.metadata) {
+      throw new Error('found no metadata');
+    }
+
+    if (new Date(rev.metadata.creationTimestamp!) > ONE_WEEK_AGO) {
+      revisionsToKeepByRecency.add(rev.metadata.name!);
+    }
+  }
+
+  console.log(
+    'revisions younger than a week count',
+    revisionsToKeepByRecency.size
+  );
+
+  // traffic tags must be cleaned up before a revision can be deleted
+  const services = ['lit-dev', 'lit-dev-playground'];
+  let serviceUpdated = false;
+  await Promise.all(
+    services.map(async (serviceName) => {
+      console.log(`fetching ${serviceName} service`);
+      const {data} = await run.projects.locations.services.get({
+        name: `projects/${PROJECT_ID}/locations/${_DEPLOY_REGION}/services/${serviceName}`,
+      });
+
+      if (!data.spec || !data.spec.traffic) {
+        throw new Error(`found no traffic for ${serviceName} service`);
+      }
+
+      console.log(`${serviceName} traffic count`, data.spec.traffic.length);
+
+      const trafficToKeep = data.spec.traffic.filter((t) => {
+        return (
+          // precence of percent indicates traffic is being routed here (i.e. prod)
+          Boolean(t.percent) ||
+          revisionsToKeepByRecency.has(t.revisionName!) ||
+          openPrs.has(parseInt(t.tag!.slice(2), 10))
+        );
+      });
+
+      console.log(`${serviceName} traffic to keep count`, trafficToKeep.length);
+
+      if (trafficToKeep.length === 0) {
+        throw new Error(`found no ${serviceName} traffic to keep`);
+      }
+
+      if (data.spec.traffic.length !== trafficToKeep.length) {
+        console.log(`updating ${serviceName} service`);
+        data.spec.traffic = trafficToKeep;
+        await run.projects.locations.services.replaceService({
+          name: `projects/${PROJECT_ID}/locations/${_DEPLOY_REGION}/services/${serviceName}`,
+          requestBody: data,
+        });
+        serviceUpdated = true;
+      }
+    })
+  );
+
+  // revision status takes some time to update
+  if (serviceUpdated) {
+    console.log('waiting 30 seconds for traffic tags to update');
+    await sleep(30_000);
```

Comment:
I found no API to actually verify if the update was done. 30 seconds is a conservative guess on my part and I haven't seen it fail with this wait. Before adding the wait, I have seen when we try to delete revisions too soon, it just fails because it still thinks it's associated with a tag. The cleanup job will fail and the next run will catch anything that wasn't deleted due to early termination.